LTE & NR Coexistance

Rashed97 opened the pull request:
This adds support for IMS routing between devices on LTE and NR.

Calls and SMS should work fine. I have tested SMS but cannot test calls due to still having QoS issues on NR, but according to the logs, it should work.

Additionally it adds some cleanup for the mysql database container. This makes it far more robust.

herlesupreeth commented:
Thank you very much for your contribution. I will try to review this MR as soon as possible.

One thing which prevents co-existence from working is I believe there should be two instances of SMF and UPF (one handling 4G and another 5G), referring to deploy-all.yaml.

herlesupreeth commented on ims_base/kamailio_init.sh:
```diff
@@ -45,7 +45,7 @@ elif [[ "$COMPONENT_NAME" =~ ^(pcscf-[[:digit:]]+$) ]]; then
 	mkdir -p /var/run/kamailio_pcscf && \
 	rm -f /kamailio_pcscf.pid && \
-	kamailio -f /etc/kamailio_pcscf/kamailio_pcscf.cfg -P /kamailio_pcscf.pid -DD -E -e
+	kamailio -M 16 -m 128 -f /etc/kamailio_pcscf/kamailio_pcscf.cfg -P /kamailio_pcscf.pid -DD -E -e
```
Can you let me know why this is needed?

Rashed97 replied:
Not entirely sure what the root issue is, but I was seeing weird/random memory freeing issues when including all 5 route configs (register, mo_N5, mo_Rx, mt_N5, mt_Rx). By excluding any one of those configs, the errors would disappear, and my researching it seemed to indicate that it was due to the process running out of memory. Bumping up the default memory to 16M/128M like this resolved the issues.

herlesupreeth commented on pcscf/kamailio_pcscf.cfg:
```diff
@@ -75,11 +75,11 @@ server_header="Server: TelcoSuite Proxy-CSCF"
 
 fork=yes
-children=4
+children=16
```
Can you let me know why this is increased?

Rashed97 replied:
I was seeing some weird errors with TCP and HTTP connections after splitting the configs, bumping this (and the TCP processes) up seemed to resolve it, tho I want to see if this was related to the memory issues noted above or not.

herlesupreeth replied:
from my experience upon increasing this value increases the memory requirement thats the reason I kept it at 4 (when compared to sample configuration)

herlesupreeth commented on pcscf/kamailio_pcscf.cfg:
```diff
@@ -75,11 +75,11 @@ server_header="Server: TelcoSuite Proxy-CSCF"
 #!ifndef TCP_PROCESSES
 # Number of TCP Processes
-#!define TCP_PROCESSES 16
+#!define TCP_PROCESSES 32
```
same as above

Rashed97 replied:
see above

herlesupreeth commented on pcscf/kamailio_pcscf.cfg:
```diff
@@ -96,7 +96,7 @@ enable_tls=yes
 #!ifndef TCP_PROCESSES
 # Number of TCP Processes
-#!define TCP_PROCESSES 3
+#!define TCP_PROCESSES 12
```
same as above

Rashed97 replied:
see above

herlesupreeth commented on pcscf/kamailio_pcscf.cfg:
```diff
@@ -226,6 +226,7 @@ loadmodule "websocket.so"
 loadmodule "cdp"
 loadmodule "cdp_avp"
 loadmodule "ims_qos"
+loadmodule "ims_diameter_server"
```
please remove this if its not used

Rashed97 replied:
Ah, sorry I thought I removed this. This was in an attempt to read the MSISDN from the Diameter Sh interface, which would have been my preferred approach vs reading from S-CSCF. I would still like to get this working, but I can remove it for now.

herlesupreeth commented on pcscf/kamailio_pcscf.cfg:
```diff
@@ -437,8 +438,19 @@ modparam("ims_qos", "include_rtcp_fd", 1)
 modparam("ims_qos", "rx_auth_expiry", UE_SUBSCRIPTION_EXPIRES)
 modparam("ims_qos", "recv_mode", 1)
 modparam("ims_qos", "dialog_direction", RX_IMS_REG_DIALOG_DIRECTION)
+# -- Diameter Server params for Sh interface --
```
please remove commented out codes. I know there are come places where there are commented out logic but just trying not to increase thats all.

Rashed97 replied:
see above

herlesupreeth commented on pcscf/kamailio_pcscf.cfg:
```diff
@@ -437,8 +438,19 @@ modparam("ims_qos", "include_rtcp_fd", 1)
 #!endif
 
+# Tables to store users subscription tech (Rx or N5)
+modparam("htable", "htable", "sub_tech=>size=4096;autoexpire=UE_SUBSCRIPTION_EXPIRES;")
```
I believe there is no UE_SUBSCRIPTION_EXPIRES defined in this branch (its in the other branch improve_xxx)

Rashed97 replied:
It should be in this branch, since I'm based on the improve_5g_ims branch which has it: https://github.com/herlesupreeth/docker_open5gs/commit/c3c907f5554b37fbbecc7a170280a9ae05e3c4c3

herlesupreeth commented on pcscf/kamailio_pcscf.cfg:
```diff
@@ -667,7 +722,50 @@ route {
+
+			# Try retrieving the IMPI from S-CSCF using the MSISDN
+			$var(msisdn_sub_id) = $ru;
+			route(GET_IMPI_FROM_SCSCF);
```
Not much of a fan of having to contact S-CSCF to fetch IMPI. Rather than this I would suggest parsing all the IMPUs present in P-Associated-URI during the time of registration and associate all of them with either NR or LTE

Rashed97 replied:
It wasn't my first approach either, I didn't want to have external connections, but it's the most reliable method I've found. At least in my research, it didn't seem that P-Associated-URI is always present, but perhaps I'm mistaken there? Maybe we could add a check against P-Associated-URI first and then fall back to S-CSCF if P-Associated-URI isn't present?

herlesupreeth replied:
> I didn't want to have external connections, but it's the most reliable method I've found. At least in my research, it didn't seem that P-Associated-URI is always present, but perhaps I'm mistaken there?

That SIP header is most of the time present (basically first value in this header sets the calling number UE needs to use)
. Say if its not present then we can always use the IMPI (IMSI based used for registration)

herlesupreeth commented on pcscf/route/register.cfg:
```diff
@@ -21,11 +21,158 @@ route[REGISTER] {
+	$var(imsi) = $tU;
+	#xlog("L_INFO", "IMSI: $var(imsi)\n");
+
+#!ifdef WITH_RX
```
please remove commented out code for features which may be added in future

Rashed97 replied:
Sorry, missed this (attempt at Diameter Sh)

herlesupreeth commented on pcscf/route/register.cfg:
```diff
@@ -21,11 +21,158 @@ route[REGISTER] {
+#!endif
+
+#!ifdef WITH_N5
+	if (!$var(msisdn)) {
```
didnt understand why this retrieving of MSISDN is needed

Rashed97 replied:
This was an attempt to fetch the MSISDN to use for checking the subscription in MO and MT routes, since the devices do not use the IMPI then, but this isn't functional on Open5GS due to missing APIs.

herlesupreeth commented on pcscf/route/register.cfg:
```diff
@@ -21,11 +21,158 @@ route[REGISTER] {
+}
+
+#!ifdef WITH_RX
+#event_route[ims_diameter_server:sh-User-Data-Answer] {
```
please remove this as well

Rashed97 commented:
> Thank you very much for your contribution. I will try to review this MR as soon as possible.
> 
> One thing which prevents co-existence from working is I believe there should be two instances of SMF and UPF (one handling 4G and another 5G), referring to deploy-all.yaml.

I've not found any issues with running both LTE and NR with a single SMF/UPF pair, but maybe I'm missing something. Everything I've tested on LTE seems to work fine but if you know what I should be on the lookout for, I can test.

Additionally, see above replied to comments

herlesupreeth commented:
Thank you for addressing the comments. Is it okay if I delay merging this MR once calling in NR is verified?

> I've not found any issues with running both LTE and NR with a single SMF/UPF pair, but maybe I'm missing something. Everything I've tested on LTE seems to work fine but if you know what I should be on the lookout for, I can test.

Thats because SMF is deployed in 4G mode when deployed using deploy-all.yaml

Rashed97 commented:
> Thank you for addressing the comments. Is it okay if I delay merging this MR once calling in NR is verified?

Sure I'm fine with that, but I personally can't do that given my continued issues with the QoS bearer issues (still not sure why it's not working even on a Threadripper 7960X with 256GB RAM). I'll upload a new chain today with some of the cleanup you requested around the commented out code, and I'll test identity retrieval from the P-Associated-URI header like you suggested as well.
 
> Thats because SMF is deployed in 4G mode when deployed using deploy-all.yaml

Are you sure? Looking at smf_init.sh, it looks like the 4G SMF config is only loaded if DEPLOY_MODE is set to 4G: https://github.com/herlesupreeth/docker_open5gs/blob/improve_5g_ims/smf/smf_init.sh#L40

When using deploy-all, its set to ALL tho, not 4G: https://github.com/herlesupreeth/docker_open5gs/blob/improve_5g_ims/deploy-all.yaml#L278

Maybe I'm missing something tho?

Rashed97 commented:
@herlesupreeth cleaned up the unused, commented out code (see above commit).

Also I added code to print the P-Associated-URI header element from the header if it's present in MO and MT and I'm not seeing it in any of the headers from my devices. Any ideas on why it's not showing?

Rashed97 commented:
@herlesupreeth I've updated the branch with some additional fixes and I've now been able to verify that calls work back and forth from an NR device to LTE (I can upload videos if you'd like verification). Not exactly sure why but my srsRAN setup suddenly stopped having the QoS PDU modification issues it was having previously.

I've also included the srsRAN update commit from the old VoNR branch, and the SGsAP commit from the main branch to bring this branch up to date.

Regarding the P-Associated-URI parsing, I still do not see that header element passed from my UEs in any of the packets received by the P-CSCF. I don't really like the S-CSCF connection either, so I'm going to look at better ways of doing this (perhaps saving it in the MySQL database or something), but as this chain stands right now, voice calls, video calls, and SMS all work between LTE and NR devices. I'm working on a new custom environment actually that may necessitate a better "tech storage" mechanism anyway: deploying an AMF, SMF, and UPF in a second location, with their own gNB, that all connect back to this main core. This will likely need a second P-CSCF at the second location as well, but still working on that.

Let me know if you'd like to see anything else or address any additional concerns with this series. Thanks!

herlesupreeth commented:
Hey!! thanks a lot for addressing all the comments. Regarding P-Associated-URI, its sent only in 200OK for SIP REGISTER. Anyways, I will take a look at the changes done recently in this PR.

Rashed97 commented:
Thanks! I’ll take a look at the 200 REGISTER message and see what I can parse there.

One more note: I’ve been playing with the processes/children again, and it looks like that change may not be needed, so I may revert that (or remove from the series). But in doing so, I’ve found something strange: the P-CSCF opens A LOT of connections to the MySQL database. Like an insane number. With the old 4 children and 16 TCP processes, it opens around 100 connections to the MySQL database. With my current changes, it’s around 350. For comparison, S-CSCF opens *one* connection and SMSC opens *two*. Any ideas on why this might be happening? It definitely doesn’t seem normal, since really it shouldn’t need that many at all, maybe 4 connections or so for usrloc, pcscf_usrloc and ims_dialog, but maybe I’m missing something.

herlesupreeth commented:
> Any ideas on why this might be happening? It definitely doesn’t seem normal, since really it shouldn’t need that many at all, maybe 4 connections or so for usrloc, pcscf_usrloc and ims_dialog, but maybe I’m missing something.

Sadly no. I havent observed that closely

Rashed97 commented:
@herlesupreeth should I reopen this against the main branch?

herlesupreeth commented:
yes please. Sorry for not being transparent here.

Rashed97 commented:
Is master where exp_5g_ims_push’s was or improve_5g_ims? What was the difference between the two branches anyway?

herlesupreeth commented:
The master branch didnt have changes in exp_5g_ims_pyhss nor the ones in improve_5g_ims. I merged changes in those branches onto master as moving forward I want to avoid having to maintain multiple branches for a feature etc.
